spectrum-view/src/versioned_avl_storage/versioned_storage.rs

```diff
@@ -0,0 +1,447 @@
+
+/// Logical serial number, used for ordering Undo operations.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct LSN(u32);
+
+const UNDO_PREFIX: &str = "p:undo";
+const DB_PREFIX: &str = "p:db";
+/// This key is associated with `Vec<VersionIdBoundaries>`, where elements are ordered from oldest
+/// to newest.
+const VERSION_ID_BOUNDARIES_PREFIX: &str = "p:bound";
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Undo<T, E> {
+    version_id: VersionId<T>,
+    key: Vec<u8>,
+    value: Vec<u8>,
+    /// E is the error type associated with the trait bound
+    /// T: TryFrom<Vec<u8>, Error = E>
+    error_phantom: PhantomData<E>,
+}
+
+impl<T, E> From<Vec<u8>> for Undo<T, E>
+where
+    T: TryFrom<Vec<u8>, Error = E>,
+    E: std::fmt::Debug,
+{
+    fn from(bytes: Vec<u8>) -> Self {
+        let version_size = bytes[0] as usize;
+        let key_size = bytes[1] as usize;
+        let value_size = bytes.len() - version_size - key_size - 2;
+        assert!(value_size > 0);
+        let version_id_bytes = bytes[2..2 + version_size].to_vec();
+        let version_id = VersionId(T::try_from(version_id_bytes).unwrap());
+
+        let key = bytes[2 + version_size..2 + version_size + key_size].to_vec();
+        let value = bytes[2 + version_size + key_size..bytes.len()].to_vec();
+        Self {
+            version_id,
+            key,
+            value,
+            error_phantom: PhantomData::default(),
+        }
+    }
+}
+
+impl<T, E> From<Undo<T, E>> for Vec<u8>
+where
+    T: Into<Vec<u8>>,
+{
+    fn from(undo: Undo<T, E>) -> Self {
+        let version_bytes: Vec<u8> = undo.version_id.0.into();
+        let version_size = version_bytes.len();
+        assert!(version_size <= 255);
+        let key_size = undo.key.len();
+        assert!(key_size <= 255);
+        let value_size = undo.value.len();
+        assert!(value_size > 0);
+        let mut bytes = Vec::with_capacity(version_size + key_size + value_size + 2);
+        bytes.push(version_size as u8);
+        bytes.push(key_size as u8);
+        bytes.extend_from_slice(&version_bytes);
+        bytes.extend_from_slice(&undo.key);
+        bytes.extend_from_slice(&undo.value);
+        bytes
+    }
+}
+
+pub struct VersionedStore<K, V, T, E> {
+    /// Storage for main data
+    db: Arc<rocksdb::OptimisticTransactionDB>,
+    key_phantom: PhantomData<K>,
+    value_phantom: PhantomData<V>,
+    version_phantom: PhantomData<T>,
+    error_conversion_phantom: PhantomData<E>,
+    num_versions_to_store: usize,
+}
+
+impl<K, V, T, E> VersionedStore<K, V, T, E>
+where
+    T: Serialize,
+{
+    pub async fn new(db_path: &str, num_versions_to_store: usize) -> Self {
+        let db = Arc::new(rocksdb::OptimisticTransactionDB::open_default(db_path).unwrap());
+        let db_cloned = db.clone();
+        spawn_blocking(move || {
+            let boundaries_key_bytes = VERSION_ID_BOUNDARIES_PREFIX.as_bytes();
+            let boundaries: Vec<VersionIdBoundaries<T>> = vec![];
```

Comment:
Yes, this array stores the first and last mapping associated with a particular version that's been committed(via the `update` method below) to storage, so that we can efficiently undo it on rollback. When we initialise a brand-new store, no versions have been committed -> empty.